src/logging.c

```diff
@@ -0,0 +1,67 @@
+#include <stdarg.h>
+#include <stdio.h>
+#include <time.h>
+
+#include "logging.h"
+
+static log_level_t min_dbg_lvl = FATAL;
+
+void log_init(const config_t* conf) {
+    min_dbg_lvl = conf->debug;
+
+    // minimum debug level cannot be higher than fatal
+    if (min_dbg_lvl > FATAL)
+        min_dbg_lvl = FATAL;
+
+    // minimum debug level cannot be lower than debug
+    if (min_dbg_lvl < DEBUG)
+        min_dbg_lvl = DEBUG;
+
+    LOG_INFO("Log level initialized to %d.", min_dbg_lvl);
+}
+
+__attribute__((format(printf, 2, 3))) void _log_str(log_level_t dbg_lvl, const char *msg, ...) {
```

Comment:
Since `_log_str` is supposed to be private/static, i.e., we only want `LOG_*` to be used, we could add the parameters `dbg_lvl_str` and `stream`. This would avoid the switch case. We should be on the safe side with this, since we either have `FATAL` by default or cap the debug level to the enums bounds.

Reply:
That's fair, will do nicely.